### ISSUE TITLE
[@typescript-eslint/no-unnecessary-type-assertion] False positives for "as const" assertions with template literals

### ISSUE THREAD

**Issue by klassm** [typescript-eslint/typescript-eslint#8737]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.2&fileType=.tsx&code=KYDwDg9gTgLgBAYwgOwM7zFCAjANsAWwClUUAVATzGAFUoBLABSmADN6Q4BeOAcgAsYMMKgBcAenHAoWKKgB0-CAWABaJMmTAEMeUgLjecAIapEKdAG4AsACgN6ONNkAhU7Qbc4AAwAkAb0wcfGJSZEpqOiYWdhAAXw0YYGQYVGkAN3oEYFRvEzMHGBtbO1BIWDgYKmA4ZmDCEhQvACVtaAATAB50BmQAcwAaOABXZABrZAgAd2QAPjgAMjh-OzhK6tEfAKrqCFY4ILwGsIiPaLYOOICe%2Bn648WuYXr6472K44tLwaHhCuAA1Yy4ejtYwwegoACiMmgmzqR1CTR4K1sax2wE2fn8zmgbjSUXu6SBILBEOQqhxUG8AzsHzsQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KAvkA&tokens=false

### Repro Code

```TypeScript
export const problemJsonTypeUriPrefix = 'https://errors.home-connect.com/';
const errorBaseUri = `${problemJsonTypeUriPrefix}contentservices` as const;

export type ProblemJson = Record<string, unknown> & {
  type: `${typeof problemJsonTypeUriPrefix}${string}/${string}`;
};

export const ValidationError: ProblemJson = {
  type: `${errorBaseUri}/validation-error`,
};
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

### tsconfig

```jsonc
{
}
```

### Expected Result

The `as const` is kept.

### Actual Result

The `as const` is marked as unnecessary.

### Additional Info

See comment [here](https://github.com/typescript-eslint/typescript-eslint/issues/8721#issuecomment-2006286737).

**Comment by SimonSchick** [comment]
Running into this issue as well, while it is technically correct for single assignments, it doesn't work in cases where the const value is then passed on since the compiler seems drop the `const`ness.

<img width="696" alt="image" src="https://github.com/typescript-eslint/typescript-eslint/assets/3259089/df7c17c0-f2dd-437a-8faf-e17afa36a836">

vs.

<img width="686" alt="image" src="https://github.com/typescript-eslint/typescript-eslint/assets/3259089/76e96c2d-3065-4839-bc35-3b88f8c079e2">

**Comment by bradzacher** [comment]
@SimonSchick as per the issue - you want https://github.com/typescript-eslint/typescript-eslint/issues/8721, not this issue.
